erpc_c/port/erpc_serial.cpp

```diff
@@ -210,18 +210,18 @@ int serial_read(int fd, char *buf, int size)
     unsigned long bread = 0;
     char temp[RX_BUF_BYTES] = { 0 };
     DWORD errors;
-    DWORD bytesToRead = 0;
+    DWORD totalBytesRead = 0;
     DWORD bytesRead = 0;
     DWORD ret = 0;
 
-    while (bytesToRead != size)
+    while (totalBytesRead != size)
     {
         do
         {
 
             ClearCommError(hCom, &errors, NULL);
 
-            if (!ReadFile(hCom, temp, RX_BUF_BYTES - bytesToRead, &bytesRead, &s_readOverlap))
+            if (!ReadFile(hCom, temp, size - totalBytesRead, &bytesRead, &s_readOverlap))
```

Comment:
It seems like what this code was trying to do is read `size` bytes and `size` may be larger than `RX_BUF_BYTES`.  When `size > RX_BUF_BYTES` it is necessary to call `ReadFile` multiple times. It seems to me like the `bytesToRead` variable is named badly.  I don't really know the details of the Windows serial API, but to me, it seems like the overall flow should be something like this:
```
char temp[RX_BUF_BYTES];

size_t totalBytesRead = 0;
while (totalBytesRead < size) {
    const DWORD bytesToRead = std::min(RX_BUF_BYTES, size - totalBytesRead);
    DWORD bytesRead;
    if (!ReadFile(hCom, temp, bytesToRead, &bytesRead, &s_readOverlap)) {
        // TODO: error handling
    }

    if (bytesRead == 0) {
        // TODO: can this happen?
    }

    memcpy(&buf[totalBytesRead], temp, bytesRead);
    totalBytesRead += bytesRead;
}

return totalBytesRead;
```

Reply:
The problem with the original implementation was that:
```
if (!ReadFile(hCom, temp, RX_BUF_BYTES - bytesToRead, &bytesRead, &s_readOverlap))
```
leads to reading all bytes which were in the RX buffer. However, erpc tries to read at first 4 bytes then should return and tries to read the remaining bytes in another call to `serial_read`. So the original implementation got stuck when it read more than 4 bytes because this loop:
```
while (bytesToRead != size)
```
was never exited due to `bytesToRead `  (naming is really confusing, so I changed it) got bigger than `size` . 

Changing the implementation to
```
if (!ReadFile(hCom, temp, size - totalBytesRead, &bytesRead, &s_readOverlap))
```
ensures that only the desired amount of data is being retrieved from the buffer.

Cosnidering you proposal:
```
if (bytesRead == 0) {
     // TODO: can this happen? 
}
```
yes this can happen, and the loop should continue until it has read the desired amount of data.
For this snippet:
```
if (!ReadFile(hCom, temp, bytesToRead, &bytesRead, &s_readOverlap)) {
     // TODO: error handling 
}
````
My aim wasn't to improve error handling and therefore would keep what is implemented, i.e.
```
{
    if (GetLastError() == ERROR_IO_PENDING)
    {
            if (!GetOverlappedResult(hCom, &s_readOverlap, &bytesRead, FALSE))
            {
                bytesRead = 0;
                totalBytesRead = 0;
            }
        }
        else
        {
            bytesRead = 0;
            totalBytesRead = 0;
        }
    }
    else
    {
        bytesRead = 0;
        totalBytesRead = 0;
    }
}
```